File: src/main/java/dev/lain/claudejb/mcp/ToonText.java
package dev.lain.claudejb.mcp;

import java.math.BigDecimal;
import java.util.regex.Pattern;

final class ToonText {

    private static final int HEX_DIGITS = 4;
    private static final String LEADING = " \t-#";
    private static final String TRAILING = " \t";
    private static final String STRUCTURAL = ":\"\\[]{}";
    private static final BigDecimal PLAIN_MIN = new BigDecimal("1e-6");
    private static final BigDecimal PLAIN_MAX = new BigDecimal("1e21");
    private static final Pattern UNQUOTED_KEY = Pattern.compile("[A-Za-z_][A-Za-z0-9_.]*");
    private static final Pattern NUMERIC_LIKE = Pattern.compile("[+-]?[0-9]+(?:[.][0-9]+)?(?:[eE][+-]?[0-9]+)?");
    private static final Pattern NUMBER = Pattern.compile("-?[0-9]+(?:[.][0-9]+)?(?:[eE][+-]?[0-9]+)?");
    private static final Pattern LEADING_ZERO = Pattern.compile("^-?0[0-9]");

    private ToonText() {
    }

    static String key(String name) {
        requireScalars(name);
        return UNQUOTED_KEY.matcher(name).matches() ? name : quote(name);
    }

    static String primitive(Object value, char delimiter) {
        if (value == null) return "null";
        if (value instanceof Boolean) return value.toString();
        if (value instanceof String) return string((String) value, delimiter);
        if (value instanceof BigDecimal) return canonical((BigDecimal) value);
        if (value instanceof Number) return canonical(new BigDecimal(value.toString()));
        throw new ToonException("not a primitive: " + value.getClass().getName());
    }

    static boolean isPrimitive(Object value) {
        return value == null || value instanceof String || value instanceof Number || value instanceof Boolean;
    }

    static String string(String value, char delimiter) {
        requireScalars(value);
        return needsQuotes(value, delimiter) ? quote(value) : value;
    }

    static String canonical(BigDecimal number) {
        BigDecimal n = number.stripTrailingZeros();
        if (n.signum() == 0) return "0";
        BigDecimal magnitude = n.abs();
        if (magnitude.compareTo(PLAIN_MIN) >= 0 && magnitude.compareTo(PLAIN_MAX) < 0) return n.toPlainString();
        String digits = n.unscaledValue().abs().toString();
        int exponent = digits.length() - 1 - n.scale();
        String mantissa = digits.length() == 1 ? digits : digits.charAt(0) + "." + digits.substring(1);
        String sign = n.signum() < 0 ? "-" : "";
        String exponentSign = exponent < 0 ? "-" : "+";
        return sign + mantissa + "e" + exponentSign + Math.abs(exponent);
    }

    static Object value(String token) {
        if (token.startsWith("\"")) return unquote(token);
        if (token.equals("true")) return Boolean.TRUE;
        if (token.equals("false")) return Boolean.FALSE;
        if (token.equals("null")) return null;
        if (NUMBER.matcher(token).matches() && !LEADING_ZERO.matcher(token).find()) return new BigDecimal(token);
        return token;
    }

    static String keyOf(String token) {
        return token.startsWith("\"") ? unquote(token) : token;
    }

    static String unquote(String token) {
        StringBuilder out = new StringBuilder();
        int i = 1;
        while (i < token.length()) {
            char c = token.charAt(i);
            if (c == '"') {
                if (i != token.length() - 1) throw new ToonException("characters after the closing quote in " + token);
                return out.toString();
            }
            if (c == '\\') i = escape(token, i + 1, out);
            else out.append(c);
            i++;
        }
        throw new ToonException("unterminated string: " + token);
    }

    private static int escape(String token, int at, StringBuilder out) {
        if (at >= token.length()) throw new ToonException("unterminated escape in " + token);
        char c = token.charAt(at);
        switch (c) {
            case '\\': out.append('\\'); return at;
            case '"': out.append('"'); return at;
            case 'n': out.append('\n'); return at;
            case 'r': out.append('\r'); return at;
            case 't': out.append('\t'); return at;
            case 'u': break;
            default: throw new ToonException("invalid escape in " + token);
        }
        int end = at + 1 + HEX_DIGITS;
        String hex = token.substring(at + 1, Math.min(end, token.length()));
        if (hex.length() != HEX_DIGITS || !hex.chars().allMatch(ToonText::isHex)) {
            throw new ToonException("truncated unicode escape in " + token);
        }
        int code = Integer.parseInt(hex, 16);
        if (Character.isSurrogate((char) code)) throw new ToonException("surrogate escape in " + token);
        out.append((char) code);
        return end - 1;
    }

    private static boolean isHex(int c) {
        return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
    }

    private static String quote(String value) {
        StringBuilder out = new StringBuilder("\"");
        for (int i = 0; i < value.length(); i++) {
            char c = value.charAt(i);
            if (c == '\\') out.append("\\\\");
            else if (c == '"') out.append("\\\"");
            else if (c == '\n') out.append("\\n");
            else if (c == '\r') out.append("\\r");
            else if (c == '\t') out.append("\\t");
            else if (c < ' ') out.append(String.format("\\u%04x", (int) c));
            else out.append(c);
        }
        return out.append('"').toString();
    }

    private static boolean needsQuotes(String s, char delimiter) {
        if (s.isEmpty() || s.equals("true") || s.equals("false") || s.equals("null")) return true;
        if (NUMERIC_LIKE.matcher(s).matches()) return true;
        if (LEADING.indexOf(s.charAt(0)) >= 0 || TRAILING.indexOf(s.charAt(s.length() - 1)) >= 0) return true;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (STRUCTURAL.indexOf(c) >= 0 || c < ' ' || c == delimiter) return true;
        }
        return false;
    }

    private static void requireScalars(String value) {
        for (int i = 0; i < value.length(); i++) {
            char c = value.charAt(i);
            boolean paired = Character.isHighSurrogate(c) && i + 1 < value.length() && Character.isLowSurrogate(value.charAt(i + 1));
            if (paired) i++;
            else if (Character.isSurrogate(c)) throw new ToonException("unpaired surrogate at index " + i);
        }
    }
}

File: src/main/kotlin/dev/lain/claudejb/model/mcp/toon/ToonText.kt
package dev.lain.claudejb.model.mcp.toon

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.JsonUnquotedLiteral
import java.math.BigDecimal
import kotlin.math.abs

internal object ToonText {

    fun key(name: String): String {
        requireScalars(name)
        return if (UNQUOTED_KEY.matches(name)) name else quote(name)
    }

    fun primitive(value: JsonPrimitive, delimiter: Char): String = when {
        value is JsonNull -> "null"
        value.isString -> string(value.content, delimiter)
        value.content == "true" || value.content == "false" -> value.content
        else -> value.content.toBigDecimalOrNull()?.let(::canonical) ?: "null"
    }

    fun string(value: String, delimiter: Char): String {
        requireScalars(value)
        return if (needsQuotes(value, delimiter)) quote(value) else value
    }

    fun canonical(number: BigDecimal): String {
        val n = number.stripTrailingZeros()
        if (n.signum() == 0) return "0"
        val magnitude = n.abs()
        if (magnitude >= PLAIN_MIN && magnitude < PLAIN_MAX) return n.toPlainString()
        val digits = n.unscaledValue().abs().toString()
        val exponent = digits.length - 1 - n.scale()
        val mantissa = if (digits.length == 1) digits else digits[0] + "." + digits.substring(1)
        val sign = if (n.signum() < 0) "-" else ""
        val exponentSign = if (exponent < 0) "-" else "+"
        return "$sign${mantissa}e$exponentSign${abs(exponent)}"
    }

    fun value(token: String): JsonElement = when {
        token.startsWith('"') -> JsonPrimitive(unquote(token))
        token == "true" -> JsonPrimitive(true)
        token == "false" -> JsonPrimitive(false)
        token == "null" -> JsonNull
        NUMBER.matches(token) && !LEADING_ZERO.containsMatchIn(token) -> number(token)
        else -> JsonPrimitive(token)
    }

    fun keyOf(token: String): String = if (token.startsWith('"')) unquote(token) else token

    fun unquote(token: String): String {
        val out = StringBuilder()
        var i = 1
        while (i < token.length) {
            val c = token[i]
            if (c == '"') {
                if (i != token.lastIndex) toonError("characters after the closing quote in $token")
                return out.toString()
            }
            i = if (c == '\\') escape(token, i + 1, out) else i.also { out.append(c) }
            i++
        }
        toonError("unterminated string: $token")
    }

    @OptIn(ExperimentalSerializationApi::class)
    private fun number(token: String): JsonPrimitive = JsonUnquotedLiteral(canonical(BigDecimal(token)))

    private fun escape(token: String, at: Int, out: StringBuilder): Int {
        val c = token.getOrNull(at) ?: toonError("unterminated escape in $token")
        ESCAPES[c]?.let {
            out.append(it)
            return at
        }
        if (c != 'u') toonError("invalid escape in $token")
        val end = at + 1 + HEX_DIGITS
        val hex = token.substring(at + 1, minOf(end, token.length))
        if (hex.length != HEX_DIGITS || !hex.all(::isHex)) toonError("truncated unicode escape in $token")
        val code = hex.toInt(16)
        if (code in SURROGATE_MIN..SURROGATE_MAX) toonError("surrogate escape in $token")
        out.append(code.toChar())
        return end - 1
    }

    private fun isHex(c: Char): Boolean = c in '0'..'9' || c in 'a'..'f' || c in 'A'..'F'

    private fun quote(value: String): String {
        val out = StringBuilder("\"")
        for (c in value) {
            when {
                c == '\\' -> out.append("\\\\")
                c == '"' -> out.append("\\\"")
                c == '\n' -> out.append("\\n")
                c == '\r' -> out.append("\\r")
                c == '\t' -> out.append("\\t")
                c < ' ' -> out.append("\\u").append(c.code.toString(16).padStart(HEX_DIGITS, '0'))
                else -> out.append(c)
            }
        }
        return out.append('"').toString()
    }

    private fun needsQuotes(s: String, delimiter: Char): Boolean =
        s.isEmpty() || s.first() in LEADING || s.last() in TRAILING || ambiguous(s, delimiter)

    private fun ambiguous(s: String, delimiter: Char): Boolean =
        s in LITERALS || NUMERIC_LIKE.matches(s) || s.any { it in STRUCTURAL || it < ' ' || it == delimiter }

    private fun requireScalars(value: String) {
        var i = 0
        while (i < value.length) {
            val c = value[i]
            val paired = c.isHighSurrogate() && i + 1 < value.length && value[i + 1].isLowSurrogate()
            if (paired) {
                i++
            } else if (c.isSurrogate()) {
                toonError("unpaired surrogate at index $i")
            }
            i++
        }
    }

    private const val HEX_DIGITS = 4
    private const val LEADING = " \t-#"
    private const val TRAILING = " \t"
    private const val STRUCTURAL = ":\"\\[]{}"
    private val LITERALS = setOf("true", "false", "null")
    private const val SURROGATE_MIN = 0xD800
    private const val SURROGATE_MAX = 0xDFFF
    private val ESCAPES = mapOf('\\' to '\\', '"' to '"', 'n' to '\n', 'r' to '\r', 't' to '\t')
    private val PLAIN_MIN = BigDecimal("1e-6")
    private val PLAIN_MAX = BigDecimal("1e21")
    private val UNQUOTED_KEY = Regex("[A-Za-z_][A-Za-z0-9_.]*")
    private val NUMERIC_LIKE = Regex("[+-]?[0-9]+(?:[.][0-9]+)?(?:[eE][+-]?[0-9]+)?")
    private val NUMBER = Regex("-?[0-9]+(?:[.][0-9]+)?(?:[eE][+-]?[0-9]+)?")
    private val LEADING_ZERO = Regex("^-?0[0-9]")
}
